### DISCUSSION TITLE
JSONPath in Grafana question

### DISCUSSION THREAD

**Discussion by m-gnaedig** [grafana/grafana-infinity-datasource#241] [Q&A] [answered]
Hello

I am still a little bit confuse.
I have to select all the values from "street_name" from the "locality": "Bonn".

If I do some tests on http://jsonpath.herokuapp.com/ it works like expected:
![JSONPath-Example](https://user-images.githubusercontent.com/30392807/151075965-70880ee5-0578-4424-a424-4ab3cfd789ac.png)

If execute some tests in Grafana than I don't get the expected result:
![JSONPath-Grafana-Example](https://user-images.githubusercontent.com/30392807/151076015-c616f975-7822-4d6b-af94-c6bc610af419.png)

Grafana Version: v8.3.3

JSON:
`[
  {
		"id": "647",
		"address": {
			"locality": "Bonn",
			"postal_code": "56867",
			"street_name": "Bahnhofstrasse",
			"street_number": 5
		},
		"something": null,
		"something_else": null
	},
  {
		"id": "648",
		"address": {
			"locality": "Bonn",
			"postal_code": "56867",
			"street_name": "Beethovenstrasse",
			"street_number": 37
		},
		"something": null,
		"something_else": null
	},
	{
		"id": "649",
		"address": {
			"locality": "Köln",
			"postal_code": "56854",
			"street_name": "Hauptstrasse",
			"street_number": 12
		},
		"something": null,
		"something_else": null
	}
]`

JSONPath:
`$.*.address[?(@.locality=='Bonn')].street_name`

I don't know if this is a Grafana bug or a Infinity bug or a problem with my JSONPath.
How can I solve this task?

Thanks in advance for every advice or idea.

**Comment by yesoreyeram** [comment, marked as answer]
Hi @m-gnaedig - Sorry for not being clear in the document. 

Infinity datasource uses [JSONPath-Plus](https://www.npmjs.com/package/jsonpath-plus) library for parsing JSONpath. According to their syntax, you need `$..*[?(@property === "locality" && @.match(/Bonn$/i))]^`. For [example](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Infinity%22,%7B%22refId%22:%22A%22,%22type%22:%22json%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22$..*%5B%3F(@property%20%3D%3D%3D%20%5C%22locality%5C%22%20%26%26%20@.match(%2FBonn$%2Fi))%5D%5E%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%5B%20%7B%20%5C%22id%5C%22:%20%5C%22647%5C%22,%20%5C%22address%5C%22:%20%7B%20%5C%22locality%5C%22:%20%5C%22Bonn%5C%22,%20%5C%22postal_code%5C%22:%20%5C%2256867%5C%22,%20%5C%22street_name%5C%22:%20%5C%22Bahnhofstrasse%5C%22,%20%5C%22street_number%5C%22:%205%20%7D,%20%5C%22something%5C%22:%20null,%20%5C%22something_else%5C%22:%20null%20%7D,%20%7B%20%5C%22id%5C%22:%20%5C%22648%5C%22,%20%5C%22address%5C%22:%20%7B%20%5C%22locality%5C%22:%20%5C%22Bonn%5C%22,%20%5C%22postal_code%5C%22:%20%5C%2256867%5C%22,%20%5C%22street_name%5C%22:%20%5C%22Beethovenstrasse%5C%22,%20%5C%22street_number%5C%22:%2037%20%7D,%20%5C%22something%5C%22:%20null,%20%5C%22something_else%5C%22:%20null%20%7D,%20%7B%20%5C%22id%5C%22:%20%5C%22649%5C%22,%20%5C%22address%5C%22:%20%7B%20%5C%22locality%5C%22:%20%5C%22K%C3%B6ln%5C%22,%20%5C%22postal_code%5C%22:%20%5C%2256854%5C%22,%20%5C%22street_name%5C%22:%20%5C%22Hauptstrasse%5C%22,%20%5C%22street_number%5C%22:%2012%20%7D,%20%5C%22something%5C%22:%20null,%20%5C%22something_else%5C%22:%20null%20%7D%20%5D%22%7D%5D),

![image](https://user-images.githubusercontent.com/153843/151141134-7ab0e38d-6cfc-414c-b588-757243ffec25.png)

![image](https://user-images.githubusercontent.com/153843/151141471-b9e771f3-6890-4611-8c82-e3bbb47c97fd.png)

In future, We may consider ["jsonpath"](https://www.npmjs.com/package/jsonpath) lib instead.


Note : 

* JSONPath-Plus [online playground](https://jsonpath-plus.github.io/JSONPath/demo/)

**Comment by m-gnaedig** [comment]
Hello Sriram

Thanks a lot for all the details and for your great project.
This is exactly what I am looking for.

Greatings 
Manfred
